### PR TITLE
ASoC: SOF: Add a prepare op to IPC topology widget ops

### PR DESCRIPTION
**Pull request by ranj063**
In order to set up a pipeline with IPC4, the total memory usage for the
pipeline needs to be calculated based on the list of connected widgets.
Add new prepare() op to struct sof_ipc_tplg_widget_ops that will be used
to calculate the moemory usage for each widget in the pipelines
associated with a PCM.

Before setting up the widgets, sof_prepare_widgets_in_path() walks
through the list of connected widgets and prepares the widgets for set
up.

Also, IPC4 requires the platform_params and the runtime PCM params in
order to prepare a widget for set up. So modify the signature of
sof_pcm_setup_connected_widgets() and sof_widget_list_setup() to accept
these are arguments.

This will need to be rebased after #3429 is merged

### CODE REVIEW

File: sound/soc/sof/pcm.c [hunk 1]
```diff
@@ -102,13 +104,13 @@ int sof_pcm_setup_connected_widgets(struct snd_sof_dev *sdev, struct snd_soc_pcm
...
+		ret = sof_widget_list_setup(sdev, spcm, platform_params, runtime_params, dir);
 		if (ret < 0) {
 			dev_err(sdev->dev, "error: failed widget list set up for pcm %d dir %d\n",
 				spcm->pcm.pcm_id, dir);
-			spcm->stream[dir].list = NULL;
-			snd_soc_dapm_dai_free_widgets(&list);
-			return ret;
+				spcm->stream[dir].list = NULL;
+				snd_soc_dapm_dai_free_widgets(&list);
+				return ret;
```
Comment (bardliao) [inline sound/soc/sof/pcm.c]: Is the indentation correct?

File: sound/soc/sof/pcm.c [hunk 2]
```diff
@@ -82,16 +82,18 @@ void snd_sof_pcm_period_elapsed(struct snd_pcm_substream *substream)
 }
 EXPORT_SYMBOL(snd_sof_pcm_period_elapsed);
 
-int sof_pcm_setup_connected_widgets(struct snd_sof_dev *sdev, struct snd_soc_pcm_runtime *rtd,
-				    struct snd_sof_pcm *spcm, int dir)
+static int
+sof_pcm_setup_connected_widgets(struct snd_sof_dev *sdev, struct snd_soc_pcm_runtime *rtd,
+				struct snd_sof_pcm *spcm,
+				struct snd_sof_platform_stream_params *platform_params,
+				struct snd_pcm_hw_params *runtime_params, int dir)
```
Comment (ujfalusi) [inline sound/soc/sof/pcm.c]: Keep it as `params` instead of using different name inside of this function (`runtime_params`, which is not that correct)?

File: sound/soc/sof/pcm.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/pcm.c]: This is no change except for the indentation increase, I guess it is an error?

File: sound/soc/sof/sof-audio.c [hunk 3]
```diff
@@ -332,10 +370,13 @@ static int sof_set_up_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_d
 }
 
 static int
-sof_setup_or_free_widgets_in_order(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget_list *list,
-				   bool dir, enum sof_widget_op op)
+sof_widget_op_in_order(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget_list *list,
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c]: `sof_widget_op_in_order` ?

File: sound/soc/sof/sof-audio.c [hunk 4]
```diff
@@ -253,6 +253,44 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+			    struct snd_pcm_hw_params *input_params, int dir)
+{
+	const struct sof_ipc_tplg_ops *ipc_tplg_ops = sdev->ipc->ops->tplg;
+	const struct sof_ipc_tplg_widget_ops *widget_ops = ipc_tplg_ops->widget;
+	struct snd_sof_widget *swidget = widget->dobj.private;
+	struct snd_soc_dapm_path *p;
+	int ret;
+
+	if (!widget_ops[widget->id].prepare || swidget->complete)
+		return 0;
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c]: and if it's sink(s) have prepare we skip those?

File: sound/soc/sof/sof-audio.c [hunk 5]
```diff
@@ -253,6 +253,44 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+	swidget->complete = true;
+
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (!p->walking && p->sink->dobj.private) {
+			p->walking = true;
+			ret = sof_prepare_widgets_in_path(sdev, p->sink, platform_params,
+							  runtime_params, input_params, dir);
+			p->walking = false;
+			if (ret < 0)
+				return ret;
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c:325]: no need to rewind?
If there is an error in the swidget's sink, should the swidget be considered complete (iow should we set swidget->complete = falsxe)?

File: sound/soc/sof/sof-audio.c [hunk 6]
```diff
@@ -350,25 +391,48 @@ sof_setup_or_free_widgets_in_order(struct snd_sof_dev *sdev, struct snd_soc_dapm
...
+			struct snd_pcm_hw_params *input_params;
+
+			/*
+			 * When walking the list of connected widgets, the input_params for each
+			 * widget is modified by the source widget in the path. Use a local
+			 * copy of the runtime params as the input_params so that the runtime
+			 * params does not get overwritten.
+			 */
+			input_params = kmemdup(runtime_params, sizeof(*runtime_params),
+					       GFP_KERNEL);
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c]: have the input_params on stack and use memcpy?

File: sound/soc/sof/sof-audio.c [hunk 7]
```diff
@@ -379,7 +443,14 @@ int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, in
...
 
-	ret = sof_setup_or_free_widgets_in_order(sdev, list, dir, SOF_WIDGET_SETUP);
+	/* prepare widgets for set up */
+	ret = sof_widget_op_in_order(sdev, list, platform_params, runtime_params,
+				     dir, SOF_WIDGET_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	ret = sof_widget_op_in_order(sdev, list, platform_params, runtime_params,
+				     dir, SOF_WIDGET_SETUP);
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c]: Ah, so the `op` might stand for OPeration?
We have so many op/ops around that it is started to get hard to know what is what...
`sof_widget_action_in_order`, which sounds a bit funny without a counterpart of `sof_widget_action_in_random / not_in_order / out_of_order`.

**Review by ujfalusi — CHANGES_REQUESTED** [review]
@ranj063, I can not really follow the change today, but I have few comments which might be valid

File: sound/soc/sof/pcm.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/pcm.c]: sorry not getting your point.It is the runtime PCM params here isnt it?

File: sound/soc/sof/sof-audio.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: no we dont. this is a recursive function.

File: sound/soc/sof/sof-audio.c (hunk 5 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:325]: no need to unwind because the widget list will be freed in case of error.

File: sound/soc/sof/sof-audio.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: widget_action with enum sof_widget_op is more odd I think

File: sound/soc/sof/sof-audio.c (hunk 5 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:325]: but I fixed the error path to clear complete

File: sound/soc/sof/sof-audio.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: ive changed it to walk_widgets_in_order()

**Comment by keqiaozhang** [comment]
SOFCI TEST

**Comment by ranj063** [comment]
@ujfalusi pleae have a look again. I've updated the error paths and added a new unprepare op to unwind.

File: sound/soc/sof/pcm.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/pcm.c]: I think @ujfalusi is saying that we usually see naming such as 
struct snd_pcm_hw_params *hw_params or
struct snd_pcm_hw_params *params

I am not sure what the 'runtime' refers to here as well.

File: sound/soc/sof/sof-audio.c [hunk 8]
```diff
@@ -253,6 +253,73 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+sink_prepare:
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (!p->walking && p->sink->dobj.private) {
+			p->walking = true;
+			ret = sof_prepare_widgets_in_path(sdev, p->sink, platform_params,
+							  runtime_params, input_params, dir);
+			p->walking = false;
+			if (ret < 0) {
+				if (!widget_ops[widget->id].unprepare)
+					widget_ops[widget->id].unprepare(swidget);
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: I don't follow why you do the unprepare here when one of the sink prepare fails, but on line 298 you don't.
This doesn't seem balanced.

File: sound/soc/sof/sof-audio.c [hunk 9]
```diff
@@ -350,25 +421,48 @@ sof_setup_or_free_widgets_in_order(struct snd_sof_dev *sdev, struct snd_soc_dapm
...
+			 * params does not get overwritten.
+			 */
+			memcpy(&input_params, runtime_params, sizeof(*runtime_params));
+
+			ret = sof_prepare_widgets_in_path(sdev, widget, platform_params,
+							  runtime_params, &input_params, dir);
+			break;
+		}
+		case SOF_WIDGET_UNPREPARE:
+			sof_unprepare_widgets_in_path(sdev, widget);
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:458]: I am not able to figure out what those states mean and what transitions are allowed.

File: sound/soc/sof/sof-audio.h [hunk 10]
```diff
@@ -38,6 +38,8 @@
 enum sof_widget_op {
 	SOF_WIDGET_FREE,
 	SOF_WIDGET_SETUP,
+	SOF_WIDGET_PREPARE,
+	SOF_WIDGET_UNPREPARE,
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.h:42]: you really ought to describe what those states mean.

It seems from reading the code that PREPARE is done before SETUP, but I have no idea how UNPREPARE/FREE are related.

File: sound/soc/sof/sof-audio.h [hunk 11]
```diff
@@ -109,6 +111,8 @@ struct sof_ipc_tplg_control_ops {
  * @token_list: List of token ID's that should be parsed for the widget
  * @token_list_size: number of elements in token_list
  * @bind_event: Function pointer for binding events to the widget
+ * @prepare: Optional op for preparing a widget for set up
+ * @unprepare: Optional op for unpreparing a widget
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.h]: can prepare be omitted and unprepare present? or is unprepare only possible when a prepare is provided?

**Review by plbossart — CHANGES_REQUESTED** [review]
I believe more documentation is needed @ranj063, the introduction the new states is far from clear to me.

File: sound/soc/sof/sof-audio.c (hunk 8 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: @plbossart this is a recursive function that goes from a source widget all the way to the sink widget in the connected widget list. So we prepare the source for every path and then recursively move on to all the sinks. If any of the sink widget's prepare() op fails, we unroll and unprepare the source leaving all widgets in the unprepared state eventually.

File: sound/soc/sof/sof-audio.c (hunk 9 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:458]: @plbossart they are not states at all. Its just a prep for sending the IPC to set up the widgets. The IPC payload for all widgets is set up in the prepare op. SO we prepare all widgets first, then set up all widgets. Conversely, we free all the widgets in the DSP and unprepare them during widget_list_free()

File: sound/soc/sof/sof-audio.h (hunk 10 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.h:42]: I've updated the commit message and add some comments. Hopefully they help.

File: sound/soc/sof/sof-audio.h (hunk 11 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.h]: you can have unprepare without a prepare but without a prepare, a widget's complete flag will never be set, so unprepare() would never ever get invoked.

File: sound/soc/sof/sof-audio.c (hunk 8 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.c]: ah yes I missed the p->sink again.

File: sound/soc/sof/sof-audio.c (hunk 9 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.c:458]: "we prepare all widgets first, then set up all widgets"

So should we check that the widgets are prepared before setting them up?
Likewise, should we check they are freed before unpreparing them?

They may be transitions, but there are implied state transitions here that may need to be checked to handle errors. Things.will.go.wrong (tm).

File: sound/soc/sof/sof-audio.h (hunk 10 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.h:42]: I would change the order to describe the logical use of these enums

enum sof_widget_op {
SOF_WIDGET_PREPARE,
 SOF_WIDGET_SETUP,
 SOF_WIDGET_FREE,
 SOF_WIDGET_UNPREPARE,
}

File: sound/soc/sof/sof-audio.h [hunk 12]
```diff
@@ -117,6 +121,11 @@ struct sof_ipc_tplg_widget_ops {
 	int token_list_size;
 	int (*bind_event)(struct snd_soc_component *scomp, struct snd_sof_widget *swidget,
 			  u16 event_type);
+	int (*prepare)(struct snd_sof_widget *swidget,
+		       struct snd_sof_platform_stream_params *platform_params,
+		       struct snd_pcm_hw_params *runtime_params,
+		       struct snd_pcm_hw_params *input_params, int dir);
+	void (*unprepare)(struct snd_sof_widget *swidget);
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.h]: ipc_prepare and ipc_unprepare for symmetry with ipc_setup and ipc_free?

File: sound/soc/sof/sof-audio.c (hunk 9 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:458]: We could check if widgets are prepared but since prepare itself is optional, im not sure it makes sense in common code. I can add this check in the IPC4 code before set up.

**Comment by plbossart** [comment]
@ranj063 pcm.c conflict?

File: sound/soc/sof/sof-audio.c (hunk 9 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:458]: but we cant check if the widget has been freed becaue we call unprepare even if the widget has not been setup, ie when setup_widgets_in_path() fails.

**Comment by ranj063** [comment]
> @ranj063 pcm.c conflict?

yes, rebased now @plbossart

File: sound/soc/sof/sof-audio.h [hunk 13]
```diff
@@ -327,6 +336,11 @@ struct snd_sof_widget {
 	struct snd_soc_component *scomp;
 	int comp_id;
 	int pipeline_id;
+	/*
+	 * complete is used to indicate that pipeline set up is complete for scheduler type
+	 * widgets. For all other types of widgets, it is used to incidcate that the widget has
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.h]: typo: indicate

File: sound/soc/sof/sof-audio.c [hunk 14]
```diff
@@ -68,6 +66,7 @@ int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 		ret = sof_widget_free(sdev, swidget->pipe_widget);
 		if (ret < 0 && !err)
 			err = ret;
+		swidget->pipe_widget->complete = 0;
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:69]: I am starting to wonder if this patch combines too many changes. this one where the meaning of 'complete' is changed, could it be a separate patch?

File: sound/soc/sof/sof-audio.h (hunk 13 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.h]: fixed

File: sound/soc/sof/sof-audio.c (hunk 14 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:69]: done

File: sound/soc/sof/sof-audio.c [hunk 15]
```diff
@@ -257,6 +257,79 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
 
+static void
+sof_unprepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget)
+{
+	const struct sof_ipc_tplg_ops *ipc_tplg_ops = sdev->ipc->ops->tplg;
+	const struct sof_ipc_tplg_widget_ops *widget_ops = ipc_tplg_ops->widget;
+	struct snd_sof_widget *swidget = widget->dobj.private;
+	struct snd_soc_dapm_path *p;
+
+	if (!widget_ops[widget->id].ipc_unprepare || !swidget->complete)
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: question: swidget->complete is always zero for non_pipeline widgets, so is the second condition not an always-true case?

Along the same lines does a pipeline widget have any sinks? I am a bit confused here.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: @plbossart this is probably the most confusing part. pipeline widget (or scheduler widgets) are not part of the connected DAPM widget list at all. So the second condition here is not al always true case. It is set to true if the prepare() is successful. And no pipeline widgets dont have any source or sinks. They dont show up in the list of connected DAPM widgets at all. That's why they are explicitly set up before a widget that's part of the pipeline gets set up.

File: sound/soc/sof/sof-audio.h (hunk 10 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-audio.h:42]: And what are these exactly?
Are they
a) `state`s
b) `operation`s (`op`s)
c) `action`s
d) something else?

File: sound/soc/sof/pcm.c [hunk 16]
```diff
@@ -84,7 +84,9 @@ EXPORT_SYMBOL(snd_sof_pcm_period_elapsed);
 
 static int
 sof_pcm_setup_connected_widgets(struct snd_sof_dev *sdev, struct snd_soc_pcm_runtime *rtd,
-				struct snd_sof_pcm *spcm, int dir)
+				struct snd_sof_pcm *spcm,
+				struct snd_sof_platform_stream_params *platform_params,
+				struct snd_pcm_hw_params *params, int dir)
```
Comment (ujfalusi) [inline sound/soc/sof/pcm.c]: nitpick: I would have the following order: sdev, rtd, spcm, params, platform_params, dir

But rtd and spcm can be derived from substream if we want to keep the parameter list in check...

File: sound/soc/sof/sof-audio.c [hunk 17]
```diff
@@ -257,6 +257,79 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+
+	/* prepare the source widget */
+	ret = widget_ops[widget->id].ipc_prepare(swidget, platform_params, runtime_params,
+					     input_params, dir);
+	if (ret < 0) {
+		dev_err(sdev->dev, "failed to prepare widget %s\n", widget->name);
+		return ret;
+	}
+
+	swidget->complete = true;
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c]: So, this is _a_ scheduler type widget if we are here, right?

Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: It is mandatory that all scheduler widget must have ipc_prepare callback set? It is needed to be able to set the complete to true...

File: sound/soc/sof/sof-audio.c [hunk 18]
```diff
@@ -257,6 +257,79 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+	struct snd_sof_widget *swidget = widget->dobj.private;
+	struct snd_soc_dapm_path *p;
+	int ret;
+
+	if (!widget_ops[widget->id].ipc_prepare || swidget->complete)
+		goto sink_prepare;
+
+	/* prepare the source widget */
+	ret = widget_ops[widget->id].ipc_prepare(swidget, platform_params, runtime_params,
+					     input_params, dir);
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c]: It is unexpected (by naming) that the `ipc_prepare` callback would modify the `input_params`, afaik it is not touching the 'runtime_params' (should this be called 'params'?).

How this works? `ipc_prepare` modifies the second params struct based on the first params _and_ what was passed via the second params? Kind of in place refine/update?
Is the `input_params` is some sort of params for the pipeline thing?

**Review by ujfalusi — CHANGES_REQUESTED** [review]
@ranj063, I'm still confused around the meaning of `sof_widget_op` and how the params are passed along via the prepare path.

Can you clarify it for me?

**Review by bardliao — DISMISSED** [review]
LGTM

File: sound/soc/sof/sof-audio.c (hunk 17 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: no @ujfalusi scheduler type widgets are not part of the list of connected DAPM widgets ar all. They do not need a prepare callback. They are prepared while walking the list of widgets in the pipeline. Please see my comment in https://github.com/thesofproject/linux/pull/3440/commits/b550ce73dd96fabc6615998bcad42f8752b6e1c1#r810449777

File: sound/soc/sof/sof-audio.c (hunk 18 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: @ujfalusi since we are walking the widgets in order from source to sink. The first widget uses the runtime_params and modifes the input_params for its sink based on what it supports (coming from topology). For example: a copier connected to a mixin, will modify the input_params to 32-bit sample size because that the only one the mixin supports

**Comment by ranj063** [comment]
@ujfalusi @bardliao have a look again. reorderd args and renamed runtime_params to just be params now

File: sound/soc/sof/sof-audio.c (hunk 17 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: @ranj063, you have added a comment to clarify what swidget->complete is in the first patch:
```c
/*
 * complete flag is used to indicate that pipeline set up is complete for scheduler type
 * widgets. It is unused for all other widget types.
 */
```
But you are setting it to true for non scheduler type of swidget?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: and the comment is modified in the next patch?

File: sound/soc/sof/sof-audio.h (hunk 10 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.h:42]: operations to perform while walking the widgets, ie set up the widgets while walking, free them or prepare them

File: sound/soc/sof/sof-audio.c (hunk 18 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: Right, but when you move to the next widget in the path, you still pass the `params` and the `input_params` in a same order.

How I see is: the first params is the _input_ params fort he given widget, the second is what will come out from the widget, so the _output_ params of the widget.
The next widget in the path would get the _output_ params of the previous widget as it's _intput_ params and produces an _output_ params. The next widget would again get the previous widget's _output_ params as it's _input_ params and produces the _output_ params.

But the code always gives the original params to all widget in the path and passing along the very same input_params struct for them to modify.

File: sound/soc/sof/sof-audio.c (hunk 17 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: I see, you also updated the comment around the complete.

File: sound/soc/sof/sof-audio.c (hunk 18 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: @ujfalusi that's one way to look at it but the reason why I named it runtime_params in the first place is this: Take this pipeline for ex: host_copier -> mixin->mixout->dai_copier. The host copier looks at the "params" to decide what its config should be. But it modifies the input_params to set the sample size to 32-bits for the mixin. Then the mixin uses the input_params to decide what its config should be. and passes the input_params unmodified to the mixout. Then mixout picks its config and passes the input_params unmodified to the dai_copier. The dai_copier does not look at the input_params, but chooses its config based on the "params". So if the params asked for 16-bits, it would do the conversion and set the input to 32-bits because thats what mixout provides and output to 16-bits because thats what the pcm params is asking for.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: so in essence, each widget gets 2 params, 1 is the unmodified PCM params and the other one is the input_params as modified by the previous widget in the path.

**Review by bardliao — DISMISSED** [review]

File: sound/soc/sof/sof-audio.c (hunk 18 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: @ranj063, thanks fir the explanation!
Basically you are passing the fw_params and a pipeline_params around and modify the pipeline_params once at the start (host_copier for playback)?
Why do you convert it back at the end?
`host_copier` -> `mixin` -> `mixout` -> `dai_copier`
`16bit` -> `32bit` -> `32bit` -> `16bit` (it gets even worst if we have 24bit)
Why not just:
`16bit` -> `32bit` -> `32bit` -> `32bit`

I know, there must be a good reason...

The naming and function of these `params` are most confusing. For the `host_copier` the first params is the _input_ or host_params or fe_params (the stream which is coming in from user space), the second params is the _output_ params of the `host_copier`, iow, what is going to be passed along the pipeline (if I understand it right).

The runtime_params might be better off to be named host or fe params?

If I get it right, this prepare will be only used by IPC4, but the 'generic' implementation is really a mirror on what IPC4 needs.
At the end you have modules with source and sink, right? you connect the sink of a module to the source of another module.
host_copier
source: host_params/fw_params
sink: host_sink_params

mixin
source: host_sink_params
sink: mixin_sink_params (basically copy of host_sink_params)

mixout
source: mixin_sink_params
sink: mixout_sink_params ( == mixin_sink_params == host_sink_params)

dai_copier
source: mixout_sink_params
sink: host_params/fw_params (? why we do another back conversion?)

Isn't this is what happening in a generic way? Should the code reflect a generic implementation instead of what IPC4 needs?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: > Why do you convert it back at the end?

@ujfalusi if the host application asks for 48k 16-bit 2ch audio, the DSP should output that isnt it? How it processes this internally shouldnt change the requirements right? @plbossart @bardliao please correct me if I am wrong.

Comment (plbossart) [reply sound/soc/sof/sof-audio.c]: @ranj063 The host application can require a format on the host side only, i.e. a PCM device. What happens in firmware and hardware is not known to the application, e.g. there could be resampling and mixing.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: @plbossart if the mixing is going to change the format to 32-bit, shoudnt there be a be_dai_link_fixup() for this case to switch the format to 32-bits as well?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: and this fixup will have to be per pipeline based on whether there's mixing happening or not. Im not sure how we can do this

Comment (plbossart) [reply sound/soc/sof/sof-audio.c]: the problem is that the mixing is not represented except in vague terms. The only place where you can influence format changes is with the DAI in the topology.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: we could set the format for the DAI copier in topology. But if it is capable of changing the format back to 16-bit, do we really need to do a fixup?

Comment (plbossart) [reply sound/soc/sof/sof-audio.c]: why would we go back to 16 bits? It's better to keep the highest possible format at the DAI level, and let the codec deal with it.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: @plbossart @ujfalusi I've changed the logic in the IPC4 patches to not convert back to 16-bits for playback but we still need the 2 params, pcm params and input_params, because for capture, the host_copier  needs to convert based on pcm params

Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: @ranj063, yes, we need to to have the host params dragged around for the capture as it applies to the closing item on the list.

Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: @ranj063, I still find the `input_params` naming a bit confusing. Yes, it is kind of that, but it is not. Especially in case of capture, it is not an input params at all.
The first params is clear: it is the host/fw/pcm params
The second is widget params, perhaps? It might be less confusing if you would not copy the host params to it, but zero it out and let the first widget to fill it up? IN that case it can be seen as the morphing param throughout the pipeline from source's sink to destination's source.
Perhaps?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: @ujfalusi why is it not input_params for capture? EVery widget has input/putput audio format irrespective of direction. So, input_params is whats being fed to a widget from the previous widget ie the prev widget's output params

Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: @ranj063, yes, we don't have disagreement on the function of the second params struct. It is just that (in my understanding) it is to get the output params of the first widget in the pipeline and from there it is going to be passed to the next one as input params and that one can modify it (I know, IPC4 is not doing that) so the next widget will be fed with the output params of the previous.
For me when you kick the `sof_unprepare_widgets_in_path()` chain to motion the second params is not input (it can not be an input on the capture case) it is just a placeholder to drag around to carry the params down the line.

This is probably my nitpicking only, if no one else have similar view, I'm fine with naming it as input_params.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: @ujfalusi if you have a suggestion for a more intuitive naming, im all ears. btw sof_unprepare_widgets_in_path() does not using params at all

Comment (plbossart) [reply sound/soc/sof/sof-audio.c]: source_params?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: source_params is good for me @ujfalusi are you OK with it too?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: @plbossart @ujfalusi renamed input_params to source_params now. Let me know what you think about this one

**Review by bardliao — DISMISSED** [review]

**Comment by ranj063** [comment]
@ujfalusi updated  now base don our discussion to use a separate flag for prepared widgets and renamed params to fe_params and pipeline_params

**Review by plbossart — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]
@ranj063, thank you for the changes!

**Review by bardliao — APPROVED** [review]
